### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by VasckMe**

### CODE REVIEW

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+//  Created by Apple Macbook Pro 13 on 13.08.22.
+//
+
+import UIKit
+// MARK: Protocol
+protocol UpdateContact {
+    func updateContact(model: Person)
+}
+
+class ContactViewController: UIViewController {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: final?

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    // MARK: Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        fillPersonData()
+    }
+    // MARK: IBAction
+    @IBAction func callButton() {
+        alert(title: "Sorry =(",
+              message: "This function can not be used, please stay with us for update!",
+              style: .alert)
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: Style:
```
alert(
         title: "Sorry =(",
         message: "This function can not be used, please stay with us for update!",
         style: .alert
)
```

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    var newModel: Person?
+    var delegate: UpdateCell?
+    
+    // MARK: Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        fillPersonData()
+    }
+    // MARK: IBAction
+    @IBAction func callButton() {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: All IBAction functions should be private

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// MARK: Protocol
+protocol UpdateContact {
+    func updateContact(model: Person)
+}
+
+class ContactViewController: UIViewController {
+
+    // MARK: IBOutlets
+    @IBOutlet private weak var phoneNumber: UILabel!
+    @IBOutlet private weak var emailAddress: UILabel!
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: Add type of UI component as suffix to all IBOutlets:
`emailAddress -> emailLabel`

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
+//
+//  ContactViewController.swift
+//  ContactsApp
+//
+//  Created by Apple Macbook Pro 13 on 13.08.22.
+//
+
+import UIKit
+// MARK: Protocol
+protocol UpdateContact {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: if this protocol is delegate, use related naming:
`UpdateContantDelegate`
or if it's a protocol use:
`UpdateContantProtocol` or `UpdateContantInterface`

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+              message: "This function can not be used, please stay with us for update!",
+              style: .alert)
+    }
+    @IBAction func editButton() {
+        performSegue(withIdentifier: "GoToEditContactVC", sender: nil)
+    }
+    
+    // MARK: Functions
+    private func fillPersonData() {
+        if let unwrapModel = model {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift:65]: use `guard`

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            title = unwrapModel.name + " " + unwrapModel.surname
+            phoneNumber.text = unwrapModel.phone
+            emailAddress.text = unwrapModel.email
+        }
+    }
+    
+    // MARK: - Navigation
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift:74]: guard

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            let editContactVC = segue.destination as? EditContactViewController,
+            let unwModel = model
+        {
+            let person = Person(name: unwModel.name,
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift:78]: style

File: ContactsApp/ContactsApp/ContactsControllers/PersonsListViewController.swift [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        header?.textLabel?.font = UIFont.boldSystemFont(ofSize: 18)
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if let cell = tableView.dequeueReusableCell(
+            withIdentifier: CustomTableViewCell.identifier,
+            for: indexPath) as? CustomTableViewCell
+        {
+            let person = DataPersons.personArray[indexPath.row]
+            if indexPath.row == 0 { cell.refresh(sender: person.email) }
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/PersonsListViewController.swift]: move closure at new line

File: ContactsApp/ContactsApp/DataPersons/DataPersons.swift [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                        "+375 29 7654321",
+                        "+48 503591179",
+                        "+375 29 6371233",
+                        "+375 29 1111233",
+                        "+375 29 8475642"]
+    
+    static var personArray: [Person] = []
+    
+    static func getRandomPerson() -> Person? {
+        if
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/DataPersons/DataPersons.swift:59]: guard

File: ContactsApp/ContactsApp/DataPersons/DataPersons.swift [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            let name = nameList.randomElement(),
+            let surname = surnameList.randomElement(),
+            let email = emails.randomElement(),
+            let phone = phones.randomElement()
+        {
+            let person = Person(name: name, surname: surname, email: email, phone: phone)
+            
+            let nameIndex = nameList.firstIndex(of: name)!
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/DataPersons/DataPersons.swift]: try to not use force unwrapping, you can move it to `guard`

File: ContactsApp/ContactsApp/Models/CustomTableViewCell.swift [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+class CustomTableViewCell: UITableViewCell {
+
+    var cellXIB = CustomCell()
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+    }
+
+    static var identifier = "CustomTableViewCell"
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/Models/CustomTableViewCell.swift]: why var?

File: ContactsApp/ContactsApp/XIB/CustomCell.swift [hunk 13]
```diff
@@ -0,0 +1,49 @@
+//
+//  CustomCell.swift
+//  ContactsApp
+//
+//  Created by Apple Macbook Pro 13 on 13.08.22.
+//
+
+import UIKit
+
+class CustomCell: UIView {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/XIB/CustomCell.swift]: final?

File: ContactsApp/ContactsApp/XIB/CustomCell.swift [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+    @IBOutlet weak var infoLabel: UILabel!
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/XIB/CustomCell.swift]: private?

File: ContactsApp/ContactsApp/XIB/CustomCell.swift [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        
+        addSubview(view)
+        view.frame = self.bounds
+        view.autoresizingMask = [.flexibleHeight, .flexibleWidth]
+        view.backgroundColor = .clear
+    }
+    
+}
+
+extension CustomCell {
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/XIB/CustomCell.swift]: make this extension for UIView, not for `CustomCell`. You can reuse this extension `fromNib func` for other cells

File: ContactsApp/ContactsApp/Models/CustomTableViewCell.swift [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    
+    //MARK: Init code
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        
+//        cellXIB.frame = contentView.bounds
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/Models/CustomTableViewCell.swift]: ???

File: ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift [hunk 17]
```diff
@@ -7,57 +7,64 @@
...
     // MARK: IBAction
-    @IBAction func callButton() {
-        alert(title: "Sorry =(",
-              message: "This function can not be used, please stay with us for update!",
-              style: .alert)
+    @IBAction private func callButton() {
+        alert(
+            title: "Sorry =(",
+            message: "This function can not be used, please stay with us for update!",
+            style: .alert)
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/ContactViewController.swift]: `)` on the next line

File: ContactsApp/ContactsApp/ContactsControllers/PersonsListViewController.swift [hunk 18]
```diff
@@ -50,9 +50,8 @@ extension PersonsListViewController: UITableViewDataSource {
             for: indexPath) as? CustomTableViewCell
         {
             let person = DataPersons.personArray[indexPath.row]
-            if indexPath.row == 0 { cell.refresh(sender: person.email) }
-            if indexPath.row == 1 { cell.refresh(sender: person.phone) }
-            cell.backgroundColor = .clear
+            if indexPath.row == 0 { cell.refresh(info: person.email) }
+            if indexPath.row == 1 { cell.refresh(info: person.phone) }
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/PersonsListViewController.swift]:
```
if indexPath.row == 1 { 
       cell.refresh(info: person.phone) 
}
```

File: ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift [hunk 19]
```diff
@@ -7,11 +7,14 @@
...
-protocol UpdateCell {
-    func updateCell(person: Person)
+protocol UpdateCellDelegate {
+    func updateCellDelegate(person: Person)
 }
 
-class TableViewController: UITableViewController {
+final class TableViewController: UITableViewController {
+    
+    let a: String? = nil
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift]: ???

File: ContactsApp/ContactsApp/ContactsControllers/EditContactViewController.swift [hunk 20]
```diff
@@ -50,7 +50,7 @@ class EditContactViewController: UIViewController {
                             email: email,
                             phone: phone)
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/EditContactViewController.swift]: style

File: ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+            let contactVC = segue.destination as? ContactViewController {
+            contactVC.model = DataPersons.personArray[indexPath.row]
+            contactVC.delegate = self
+        }
+    }
+}
+
+extension TableViewController: UpdateCellDelegate {
+    func updateCellDelegate(person: Person) {
+        if
```
Comment (KoalisMe) [inline ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift]: guard

File: ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+        super.viewDidLoad()
+        DataPersons.fillPersonArray()
+        self.navigationItem.rightBarButtonItem = self.editButtonItem
+    }
+
+    // MARK: - Table view data source
+
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
```
Comment (progmev) [inline ContactsApp/ContactsApp/ContactsControllers/TableViewController.swift]: можно не писать)

**Review by progmev — APPROVED** [review]
Отлично)))

**Pull request merged**